src/main/java/build/buildfarm/worker/ReportResultStage.java

```diff
@@ -0,0 +1,182 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package build.buildfarm.worker;
+
+import build.buildfarm.common.Digests;
+import build.buildfarm.instance.Instance;
+import build.buildfarm.v1test.CASInsertionControl;
+import com.google.common.collect.ImmutableList;
+import com.google.devtools.remoteexecution.v1test.Action;
+import com.google.devtools.remoteexecution.v1test.ActionResult;
+import com.google.devtools.remoteexecution.v1test.Digest;
+import com.google.devtools.remoteexecution.v1test.ExecuteOperationMetadata;
+import com.google.devtools.remoteexecution.v1test.ExecuteResponse;
+import com.google.devtools.remoteexecution.v1test.OutputFile;
+import com.google.longrunning.Operation;
+import com.google.protobuf.Any;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.InvalidProtocolBufferException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+class ReportResultStage extends AbstractPipelineStage {
+  private final Worker worker;
+
+  public static class NullStage implements PipelineStage {
+    @Override
+    public boolean claim() { return true; }
+    @Override
+    public void release() { }
+    @Override
+    public OperationContext take() { throw new UnsupportedOperationException(); }
+    @Override
+    public void offer(OperationContext operation) { }
+    @Override
+    public void setInput(PipelineStage input) { }
+    @Override
+    public void run() { }
+    @Override
+    public void close() { }
+    @Override
+    public boolean isClosed() { return false; }
+  }
+
+  ReportResultStage(Worker worker, PipelineStage error) {
+    super(new NullStage(), error);
+    this.worker = worker;
+  }
+
+  @Override
+  public void run() {
+    while (!isClosed() || claimed) {
+      iterate();
+    }
+    close();
+  }
+
+  @Override
+  protected OperationContext tick(OperationContext operationContext) {
+    final String operationName = operationContext.operation.getName();
+    Poller poller = new Poller(worker.config.getOperationPollPeriod(), () -> {
+          boolean success = worker.instance.pollOperation(
+              operationName,
+              ExecuteOperationMetadata.Stage.EXECUTING);
+          return success;
+        });
+    new Thread(poller).start();
+
+    ActionResult.Builder resultBuilder;
+    try {
+      resultBuilder = operationContext
+          .operation.getResponse().unpack(ExecuteResponse.class).getResult().toBuilder();
+    } catch (InvalidProtocolBufferException ex) {
+      poller.stop();
+      return null;
+    }
+
+    ImmutableList.Builder<ByteString> contents = new ImmutableList.Builder<>();
+    if (resultBuilder.getExitCode() == 0) {
```

Comment:
Wait, are you only uploading artifacts to CAS on success? Then, how will the user see logs of failed tests, for example?

Reply:
I was. This appears to be incorrect as we've discussed offline: regardless of exit status, all output files (that exist) should be read from and uploaded into the CAS at the completion of an action execution. Will adjust